File: console/src/pages/policy.ts
import { getPolicy, reloadPolicy } from "../api";
import { h, mount } from "../dom";
import { fmtIso, fmtUsd } from "../format";
import { toast, toastError } from "../toast";
import type { ConfigRoute, PersistedRoute, SplitEntry } from "../types";
import { kv, openKavachModeInfo, th } from "../ui";

export async function renderPolicy(target: HTMLElement, signal: AbortSignal): Promise<void> {
  const headerInfo = h("div", { class: "card" });
  const routesBody = h("tbody");
  const pricingBody = h("tbody");

  const layout = h("div", {}, [
    h("div", { class: "page-header" }, [
      h("h1", {}, "Policy"),
      h("div", { class: "controls" }, [
        h("button", { class: "ghost", events: { click: () => refresh() } }, "Refresh"),
        h("button", {
          class: "primary",
          events: {
            click: async () => {
              try {
                const r = await reloadPolicy(signal);
                toast(
                  `Reloaded: ${r.config_routes} config + ${r.stored_routes} stored routes, ${r.pricing_entries} pricing entries`,
                  "ok",
                );
                await refresh();
              } catch (e) {
                toastError("Reload failed", e);
              }
            },
          },
        }, "Reload policy"),
      ]),
    ]),
    headerInfo,
    h("h3", { style: { marginTop: "16px", fontSize: "14px" } }, "Effective routes"),
    h("p", { class: "help-block" }, "Config routes evaluate first (in declaration order), then persisted routes (by position)."),
    h("div", { class: "table-wrap" }, [
      h("table", {}, [
        h("thead", {}, h("tr", {}, [
          th("Source"),
          th("Order"),
          th("Match"),
          th("Primary / split"),
          th("Fallbacks"),
        ])),
        routesBody,
      ]),
    ]),
    h("h3", { style: { marginTop: "24px", fontSize: "14px" } }, "Pricing table"),
    h("div", { class: "table-wrap" }, [
      h("table", {}, [
        h("thead", {}, h("tr", {}, [
          th("Model"),
          th("Input USD / 1K"),
          th("Output USD / 1K"),
        ])),
        pricingBody,
      ]),
    ]),
  ]);
  mount(target, layout);

  async function refresh(): Promise<void> {
    try {
      const pol = await getPolicy(signal);
      const k = pol.kavach;
      const driftSummary = k.drift.enabled
        ? `${k.drift.detectors.length} active (warn threshold ${k.drift.warning_threshold})`
        : "disabled";
      const driftDetail = k.drift.detectors
        .map((d) => `${d.name} ${JSON.stringify(d.parameters)}`)
        .join("; ") || "(none)";
      headerInfo.replaceChildren(
        h("h3", { style: { marginTop: 0, fontSize: "14px" } }, "Loaded from"),
        kv([
          ["Config path", pol.config_path],
          ["Default provider", pol.default_provider ?? "(first registered)"],
          ["Registered providers", pol.providers.join(", ") || "(none)"],
          ["Last refreshed", fmtIso(new Date().toISOString())],
        ]),
        h("h3", { style: { marginTop: "12px", fontSize: "14px" } }, [
          "Kavach ",
          h("span", {
            class: `badge ${k.mode === "enforce" ? "ok" : "code"}`,
            title: k.mode === "observe"
              ? "Observe mode: every verdict is logged but nothing is blocked. Click for the recipe to go live."
              : k.mode === "enforce"
                ? "Enforce mode: default-deny is active. Click for details."
                : "Click for mode details.",
            style: { cursor: "pointer" },
            events: { click: () => openKavachModeInfo(k.mode) },
          }, k.mode),
        ]),
        kv([
          ["Policy path", k.policy_path ?? "(inline in marg.toml)"],
          ["Policy hash", k.policy_source_hash],
          ["Loaded at", fmtIso(k.loaded_at)],
          ["Policy rules", String(k.policy_rule_count)],
          ["Invariants", String(k.invariant_count)],
          ["Audit chain length", String(k.audit_chain_length)],
          ["Chain head hash", k.audit_chain_head_hash],
          ["Permit signer", `${k.permit_signer.enabled ? "on" : "off"} (${k.permit_signer.algorithm}, key ${k.permit_signer.key_id})`],
          ["Drift detection", driftSummary],
          ["Drift detectors", driftDetail],
          ["kavach-core", k.core_version],
          ["kavach-pq", k.pq_version],
        ]),
      );

      routesBody.replaceChildren();
      const allRows: Array<{ source: string; idx: number; row: HTMLElement }> = [];
      pol.config_routes.forEach((r, i) => {
        allRows.push({ source: "config", idx: i, row: routeRow("config", i + 1, r as ConfigRoute) });
      });
      pol.stored_routes.forEach((r, i) => {
        allRows.push({ source: "stored", idx: i, row: routeRow("stored", r.position ?? i + 1, r as PersistedRoute) });
      });
      if (allRows.length === 0) {
        routesBody.appendChild(h("tr", {}, h("td", { colspan: 5, class: "empty" }, "No routes configured.")));
      } else {
        for (const r of allRows) routesBody.appendChild(r.row);
      }

      pricingBody.replaceChildren();
      if (pol.pricing.length === 0) {
        pricingBody.appendChild(h("tr", {}, h("td", { colspan: 3, class: "empty" }, "Using built-in pricing defaults.")));
      } else {
        for (const p of pol.pricing) {
          pricingBody.appendChild(h("tr", {}, [
            h("td", {}, h("code", {}, p.model)),
            h("td", {}, fmtUsd(p.input_per_1k_usd)),
            h("td", {}, fmtUsd(p.output_per_1k_usd)),
          ]));
        }
      }
    } catch (e) {
      if ((e as { name?: string }).name === "AbortError") return;
      toastError("Failed to load policy", e);
    }
  }

  function routeRow(source: "config" | "stored", order: number, r: ConfigRoute | PersistedRoute): HTMLElement {
    const isStored = source === "stored";
    const matchModel = isStored ? (r as PersistedRoute).match_model : ((r as ConfigRoute).match?.model ?? null);
    const matchTeam = isStored ? (r as PersistedRoute).match_team : ((r as ConfigRoute).match?.team ?? null);
    const primary = isStored ? (r as PersistedRoute).primary : ((r as ConfigRoute).primary ?? null);
    const primaryModel = isStored ? (r as PersistedRoute).primary_model : null;
    const fallbacks = isStored ? ((r as PersistedRoute).fallbacks ?? []) : ((r as ConfigRoute).fallback ?? []);
    const split = isStored ? ((r as PersistedRoute).split ?? []) : ((r as ConfigRoute).split ?? []);
    return h("tr", {}, [
      h("td", {}, h("span", { class: `badge ${isStored ? "code" : "ok"}` }, source)),
      h("td", {}, String(order)),
      h("td", {}, matchSummary(matchModel ?? null, matchTeam ?? null)),
      h("td", {}, primarySummary(primary ?? null, primaryModel ?? null, split)),
      h("td", {}, fallbacks.join(", ") || "-"),
    ]);
  }

  await refresh();
}

function matchSummary(model: string | null, team: string | null): string {
  const parts: string[] = [];
  if (model) parts.push(`model=${model}`);
  if (team) parts.push(`team=${team}`);
  return parts.join(" ") || "(any)";
}

function primarySummary(
  primary: string | null,
  primaryModel: string | null,
  split: SplitEntry[],
): string {
  if (split.length > 0) {
    return `split: ${split.map((s) => `${s.provider}${s.model ? `:${s.model}` : ""}@${s.weight}`).join(", ")}`;
  }
  if (primary) {
    return `primary: ${primary}${primaryModel ? `:${primaryModel}` : ""}`;
  }
  return "-";
}

File: marg-server/src/admin/handlers/policy.rs
use axum::extract::State;
use axum::Json;
use serde_json::{json, Value};

use marg_core::Config;

use crate::admin::error::AdminError;
use crate::kavach::emit_policy_reload;
use crate::policy;
use crate::state::AppState;

pub async fn view(State(state): State<AppState>) -> Result<Json<Value>, AdminError> {
    let cfg = Config::load(state.config_path.as_str())
        .map_err(|e| AdminError::Internal(format!("read config: {}", e)))?;
    let stored = state.storage.list_routes().await?;
    let providers: Vec<String> = state.providers.keys().cloned().collect();
    let mode = state.kavach.mode.load_full();
    let policy_path = state
        .kavach
        .policy_source_path
        .as_ref()
        .map(|p| p.display().to_string());
    let policy_hash = state.kavach.policy_source_hash.load_full();
    let loaded_at = state.kavach.policy_loaded_at.load_full();
    let signer = &state.kavach.permit_signer;
    let drift = state.kavach.drift_state.load_full();
    let drift_detectors_json: Vec<Value> = drift
        .detectors
        .iter()
        .map(|d| json!({ "name": d.name, "parameters": d.parameters }))
        .collect();
    Ok(Json(json!({
        "config_path": state.config_path.as_str(),
        "providers": providers,
        "default_provider": cfg.providers.default,
        "config_routes": cfg.routes,
        "stored_routes": stored,
        "pricing": cfg.pricing,
        "kavach": {
            "mode": mode.as_str(),
            "policy_path": policy_path,
            "policy_source_hash": policy_hash.as_str(),
            "loaded_at": loaded_at.as_str(),
            "policy_rule_count": *state.kavach.policy_rule_count.load_full(),
            "invariant_count": *state.kavach.invariant_count.load_full(),
            "audit_chain_length": state.kavach.audit_chain.len(),
            "audit_chain_head_hash": state.kavach.audit_chain.head_hash(),
            "core_version": crate::KAVACH_CORE_VERSION,
            "pq_version": crate::KAVACH_PQ_VERSION,
            "permit_signer": {
                "enabled": signer.enabled,
                "algorithm": signer.algorithm,
                "key_id": signer.key_id,
            },
            "drift": {
                "enabled": drift.enabled,
                "warning_threshold": drift.warning_threshold,
                "detectors": drift_detectors_json,
            },
        },
    })))
}

pub async fn reload(State(state): State<AppState>) -> Result<Json<Value>, AdminError> {
    let previous_hash = state.kavach.policy_source_hash.load_full().as_str().to_string();
    match policy::reload(&state).await {
        Ok(outcome) => {
            // Record the reload in the signed audit chain.
            let source_path = outcome.kavach_source_path.as_ref().map(std::path::PathBuf::from);
            emit_policy_reload(
                &state.kavach.audit_chain,
                "admin",
                &outcome.kavach_previous_hash,
                &outcome.kavach_new_hash,
                source_path.as_ref(),
                outcome.kavach_rules,
                outcome.kavach_invariants,
                true,
                None,
            );
            Ok(Json(json!({
                "reloaded": true,
                "config_routes": outcome.config_routes,
                "stored_routes": outcome.stored_routes,
                "pricing_entries": outcome.pricing_entries,
                "kavach": {
                    "mode": outcome.kavach_mode,
                    "previous_hash": outcome.kavach_previous_hash,
                    "new_hash": outcome.kavach_new_hash,
                    "policy_rule_count": outcome.kavach_rules,
                    "invariant_count": outcome.kavach_invariants,
                    "source_path": outcome.kavach_source_path,
                },
            })))
        }
        Err(err) => {
            emit_policy_reload(
                &state.kavach.audit_chain,
                "admin",
                &previous_hash,
                &previous_hash,
                None,
                0,
                0,
                false,
                Some(err.to_string()),
            );
            Err(AdminError::BadRequest(err.to_string()))
        }
    }
}
